File: dsa/word-break/ts/word_break.ts
const log: Console["log"] = console.log.bind(console);

export function wordBreak(str: string, words: Array<string>): boolean {
  const n: number = str.length;
  let maxLen: number = 0;

  for (const word of words) {
    maxLen = Math.max(maxLen, word.length);
  }

  const memo: Array<boolean> = Array(n + 1).fill(false);
  memo[0] = true;

  for (let i = 1; i <= n; ++i) {
    for (let j = i - 1; j >= Math.max(0, i - maxLen); --j) {
      const seg: string = str.substring(j, i);
      if (memo[j] && words.includes(seg)) {
        memo[i] = true;
        break;
      }
    }
  }

  return memo[n];
}

File: dsa/word-break/py/wordbreak.py
def word_break(s: str, words: set[str]) -> bool:
    n = len(s)
    max_len = max(len(word) for word in words) if words else 0

    memo = [False] * (n + 1)
    memo[0] = True

    for i in range(1, n + 1):
        ini = max(0, i - max_len)
        for j in range(i - 1, ini - 1, -1):
            seg = s[j:i]
            if memo[j] and seg in words:
                memo[i] = True
                break

    return memo[n]
